core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -97,14 +97,19 @@
  */
 @ThreadSafe
 public class IndexedSet<T> implements Iterable<T> {
-  /** All objects in the set. */
-  private final Set<T> mObjects = new HashSet<T>();
+  /**
+   * All objects in the set.
+   *
+   * When an object is added, it is first added to mObjects and then added to the indices;
+   * when an object is removed, it is first removed from the indices and finally removed from
+   * mObjects. This order is necessary so that if an object can be added to mObjects, it is sure
+   * that it does not exist in indices.
+   */
+  private final Set<T> mObjects = CommonUtils.newConcurrentHashSet();
```

Comment:
why do we need mObjects? if it is just to provide for a convenient iterator, I think it should be possible to implement the iterator using one of the indices, which would remove the need for carefully ordering mObjects and index operations